source/_posts/2015-11-16-immutable-strings-in-ruby-2-dot-3.markdown

```diff
@@ -0,0 +1,136 @@
+
+* Equality comparison affected
+* Better performance
+* Thread safety
+
+
+Currently strings immutability is included as opt-in, as the first step of a multi
+stage rollout and if it works well it will be [enabled by default on Ruby 3.0][2].
+So it still can end up being just an experiment. Time will tell.
+
+<!--more-->
+
+# Immutability
+
+Quoting [wikipedia][3]:
+
+> An immutable object is an object whose state cannot be modified after it is created.
+
+As immutable objects state cannot be changed it doesn't make sense to copy the
+same object over and over. Following that approach it will have three major
+consequences:
+
+* Data duplication
+* Unnecessary object allocation
+* Thread unsafe
+
+All of this can be mitigated using references to objects. Allowing the interpreter
+to copy a reference to an object instead copying the object itself. This can have
+certain impact on performance when working with a large codebase.
+
+
+# Moving towards immutability
+
+On this initial phase of Ruby's strings immutability, we can enable this feature
+for the whole project through a flag or on a per file basis using a pragma. Furthermore
+for a more detailed error stack while debugging another flag was added.
+
+If you are working on an existent project and want to introduce this new feature,
+obviously the most appropriate method is to use the pragma approach. Just add the
+following comment at the beginning of each file you want to switch to immutable
+strings and your done. Well, not really. Probably you will have some errors to fix.
+
+    # frozen_string_literal: true
+
+But, if today you woken up feeling like a kamikaze enable frozen strings for the
+whole project and prepare yourself for some serious bug hunting. So, to enable
+string immutability on a project basic add the following flag when running your
+app.
+
+    --enable-frozen-string-literal
+
+One way or another, its quite probable that you will encounter some bugs.
+
+Lets run a quick example, create a file named `replace.rb` with the following lines
+
+{% codeblock lang:ruby %}
+  str = "abcde"
+  str.gsub!("c", "b")
+{% endcodeblock %}
+
+If we run `ruby --enable-frozen-string-literal replace.rb` an error like the one
+below is thrown.
+
+    replace.rb:2:in `gsub!': can't modify frozen String (RuntimeError)
+            from replace.rb:2:in `<main>'
+
+For a small example fixing the error is straightforward, but what if we are on large
+codebase trying to modify a string has been passed between several methods distributed
+on multiple files? Finding the original string can be cumbersome. Thats why a debug
+flag was added, if we the previous sample adding `--enable-frozen-string-literal-debug`
+
+    replace.rb:2:in `gsub!': can't modify frozen String, created at replace.rb:1 (RuntimeError)
+      from replace.rb:2:in `<main>'
+
+As you can see a subtle change is made to the error output, now we have exact
+information about were the string was created.
+
+ ### Equality
```

Comment:
The `###` is not converting the `Equality` into a title. Maybe it's just a title like the others?